File: services/grpc-java/src/main/java/com/omnitest/apijava/infrastructure/telemetry/RpcMetricsInterceptor.java
// ============================================================================
// File: infrastructure/telemetry/RpcMetricsInterceptor.java
// Purpose: Time every unary RPC as rpc.server.duration (ms) for Grafana RED.
//          Collector namespace opn → opn_rpc_server_duration_milliseconds_*.
// SOLID: SRP — interceptor only. No-op when OpenTelemetry is noop.
// ============================================================================

package com.omnitest.apijava.infrastructure.telemetry;

import io.grpc.ForwardingServerCall;
import io.grpc.Metadata;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;
import io.grpc.Status;
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.metrics.DoubleHistogram;
import net.devh.boot.grpc.server.interceptor.GrpcGlobalServerInterceptor;
import org.springframework.stereotype.Component;

/**
 * Global gRPC interceptor. Records duration even when the SDK is no-op (then discarded).
 */
@Component
@GrpcGlobalServerInterceptor
public class RpcMetricsInterceptor implements ServerInterceptor {

    private static final AttributeKey<String> STATUS = AttributeKey.stringKey("rpc.grpc.status_code");
    private static final AttributeKey<String> SERVICE = AttributeKey.stringKey("service.name");

    private final DoubleHistogram duration = GlobalOpenTelemetry.getMeter("grpc-java")
            .histogramBuilder("rpc.server.duration")
            .setUnit("ms")
            .build();

    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
            ServerCall<ReqT, RespT> call,
            Metadata headers,
            ServerCallHandler<ReqT, RespT> next) {
        long started = System.nanoTime();
        ServerCall<ReqT, RespT> wrapping = new ForwardingServerCall.SimpleForwardingServerCall<>(call) {
            @Override
            public void close(Status status, Metadata trailers) {
                double ms = (System.nanoTime() - started) / 1_000_000.0;
                String code = status == null ? "2" : String.valueOf(status.getCode().value());
                duration.record(ms, Attributes.of(STATUS, code, SERVICE, "grpc-java"));
                super.close(status, trailers);
            }
        };
        return next.startCall(wrapping, headers);
    }
}

File: services/grpc-csharp/Infrastructure/Telemetry/RpcMetricsInterceptor.cs
// ==============================================================================
// File: Infrastructure/Telemetry/RpcMetricsInterceptor.cs
// Purpose: Time every unary RPC as rpc.server.duration (ms) for Grafana RED.
// SOLID: SRP — interceptor only. Meter is a no-op unless OTEL SDK is registered.
// ==============================================================================

using Grpc.Core;
using Grpc.Core.Interceptors;
using System.Diagnostics;
using System.Diagnostics.Metrics;

namespace OmniTest.Polyglot.Nexus.Api.CSharp.Infrastructure.Telemetry;

/// <summary>Records rpc.server.duration for each inbound gRPC call.</summary>
public sealed class RpcMetricsInterceptor : Interceptor
{
    private static readonly Meter Meter = new(OtelSetup.MeterName);
    private static readonly Histogram<double> Duration = Meter.CreateHistogram<double>(
        OtelSetup.DurationInstrument, unit: "ms");

    public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
        TRequest request,
        ServerCallContext context,
        UnaryServerMethod<TRequest, TResponse> continuation)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            return await continuation(request, context);
        }
        finally
        {
            var code = context.Status.StatusCode == StatusCode.OK ? "0" : ((int)context.Status.StatusCode).ToString();
            Duration.Record(sw.Elapsed.TotalMilliseconds, new TagList
            {
                { "rpc.grpc.status_code", code },
                { "service.name", "grpc-csharp" }
            });
        }
    }
}
